Set lifecycle rules

Lowercases opened the pull request:
As a test for r101-api.

Lowercases committed:
Set lifecycle rules

isobit commented on .circleci/config.yml:
```diff
@@ -32,6 +32,36 @@ workflows:
           tag: "latest,$CIRCLE_BRANCH,$CIRCLE_SHA1"
           # create the AWS ECR repo if it does not exist already.
           create-repo: true
+          lifecycle-policy-path: >
```
The documentation says this should be a path, not the actual content.

Lowercases replied:
Yes! But it's a lie.

I started adding a file as in https://github.com/remind101/r101-api/pull/16623/files#diff-2c27a3138e2c83cb750d9a8a3b415799e8f3806ae6bd8a9d06a6abb23e018d36 (only left there for dev purposes) only to get `Invalid length for parameter lifecyclePolicyText, value: 28, valid min length: 100` (https://app.circleci.com/pipelines/github/remind101/r101-api/11123/workflows/b3566098-95c6-4f13-9fd7-8a7f3f79a4d1/jobs/257499).

After my initial wtf moment, my suspects were confirmed when I tried the current approach and bingo! got `An error occurred (AccessDeniedException) when calling the PutLifecyclePolicy operation: User: arn:aws:iam::************:user/erc-stage-circleci is not authorized to perform: ecr:PutLifecyclePolicy on resource: arn:aws:ecr:*********:************:repository/remind101/r101-api because no identity-based policy allows the ecr:PutLifecyclePolicy action` (https://app.circleci.com/pipelines/github/remind101/r101-api/11188/workflows/00a85887-5e28-4912-81d6-94738f99dbe9/jobs/258057).

Trust no one.

isobit replied:
WTF lol. Well, at least it would be good to have a comment to explain that.

> valid min length: 100

:thinking:

Lowercases replied:
What do you mean with "at least"?

Lowercases replied:
added comment

Lowercases committed:
Set a comment

isobit commented on .circleci/config.yml:
```diff
@@ -32,6 +32,36 @@ workflows:
           # create the AWS ECR repo if it does not exist already.
           create-repo: true
+          lifecycle-policy-path: >
+            {
+              "rules": [
+                {
+                  "rulePriority": 1,
+                  "description": "Keep latest master image",
+                  "selection": {
+                    "tagStatus": "tagged",
+                    "tagPrefixList": ["master", "main"],
+                    "countType": "imageCountMoreThan",
+                    "countNumber": 5
+                  },
+                  "action": {
+                    "type": "expire"
+                  }
+                },
```
IIUC, this is super dangerous; because we re-tag the `master` image on any CI run of `master`, the image that's actually being used in prod will get marked for expiration immediately after any build if that prod image is >90 days old. This is especially bad if for example, a PR gets merged, build, and pushed to the `master` tag, but flaky tests prevent it from being deployed. If it sits in that state, the in-use prod image will expire and we'll be unable to start new tasks.

Lowercases replied:
These are valid concerns -- after some discussion we've thought about mitigating these by tagging all images with $CIRCLE_BRANCH-$CIRCLE_BUILD_NUM, which should make the first rule protect any such image.

We've also increase the kept images to 8,000, i.e. 80% of the AWS ECR limit.

Lowercases committed:
Guard master and main from ECR lifecycle rules

Avoid removing "master" and "main" tagged images to avoid removing a
running image after aggressive master/main rebuilding.

In order to glob all master and main images, add a composite tag for the
branch and build number.

isobit commented on .circleci/config.yml:
```diff
@@ -28,12 +28,17 @@ workflows:
+          # tags for this image, comma separated. A composite of branch+sha1 is
+          # added so the lifecycle rule (below) can filter images beginning
+          # with "master" or "main" to avoid applying aggressive rules.
+          tag: "latest,$CIRCLE_BRANCH,$CIRCLE_SHA1,$CIRCLE_BRANCH-$CIRCLE_BUILD_NUM"
```
I think I'd still want to keep the commit SHA in the prefixed tag, because it's nice to have that info embedded and also would make it easier to transition to something other than CircleCI without causing conflicts.

rleira reviewed (approved):
Looks great I think that having a space of 10k and making the master/main branches have a longer longevity (if I understood that part) covers our cases very well. Custom branches for custom deploys will probably be due well before 90 days
